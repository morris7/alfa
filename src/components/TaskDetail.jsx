import React, { Component } from 'react';
import { Link } from 'react-router-dom';
import styled from 'styled-components';
import Message from '../assets/note.svg';
import Phone from '../assets/phone.svg';
import Button from './Button';

class TaskDetail extends Component {
  constructor(props) {
    super(props);

    this.state = {
      isLoaded: false
    };
  }

  componentDidMount() {
    const { id } = this.props.match.params;

    fetch(`http://localhost:3001/posts/${id}`)
      .then(res => res.json())
      .then(
        result => {
          this.setState({
            result,
            isLoaded: true
          });
        },
        error => {
          this.setState({
            isLoaded: true,
            error
          });
        }
      );
  }

  render() {
    const { error, isLoaded, result } = this.state;

    if (isLoaded && !error) {
      return (
        <Article>
          <Section className="taskInfo">
            <h2>Task information</h2>
            <hr />
            <p>{result.taskBody}</p>
          </Section>

          <Section className="logInfo">
            <h2>Log </h2>
            <hr />
            <ul>
              {result.log.map((log, index) => {
                const Type = log.type == 'phone' ? Phone : Message;
                return (
                  <li key={index}>
                    <img src={Type} />
                    <span>{log.content}</span>
                  </li>
                );
              })}
            </ul>
            <Button text="Add Call" />
            <Button text="Add Note" />
          </Section>
          <Link to="/">Back</Link>
        </Article>
      );
    } else {
      return null;
    }
  }
}

const Section = styled.section`
  background: #ffffff;
  box-shadow: 0 4px 4px 0 rgba(0, 0, 0, 0.26), 0 7px 14px 0 rgba(0, 0, 0, 0.19);
  border-radius: 2px;
  width: 100%;
  min-height: 200px;
  padding: 15px;
  margin-bottom: 25px;

  li {
    list-style: none;
    padding: 10px 0px;
    img {
      padding-right: 10px;
    }
    &:nth-child(even) {
      background: #f2f2f2;
    }
  }
`;

const Article = styled.article`
  a {
    text-decoration: none;
  }
`;

export default TaskDetail;
