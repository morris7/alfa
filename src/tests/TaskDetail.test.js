import React from 'react';
import { mount } from 'enzyme';
import toJson from 'enzyme-to-json';
import 'jest-styled-components';
import { MemoryRouter } from 'react-router';

import TaskDetail from '../components/TaskDetail';
import { createMockFetchSuccess, createMockFetchFailure } from './testUtils';

const FETCH_URL = 'http://localhost:3001/posts/1';
const mockRouter = {
  match: {
    params: { id: 1 }
  }
};
const taskDetail = {
  id: '1',
  taskNumber: '38263',
  date: '19/09/2017 13:41',
  customer: 'Freida Bentley',
  taskBody: 'Task body',
  note: 'Useful note',
  log: [
    {
      type: 'phone',
      content: 'content'
    }
  ]
};
const successResponse = {
  result: { ...taskDetail },
  isLoaded: true
};
const failResponse = { isLoaded: true, error: true };

let props;

describe('when rendering a TaskDetail component', () => {
  let wrapper;

  beforeEach(() => {
    const allProps = { ...props, ...mockRouter };

    wrapper = mount(
      <MemoryRouter>
        <TaskDetail {...allProps} />
      </MemoryRouter>
    );
  });

  describe('when the fetch fails', () => {
    beforeAll(() => {
      jest
        .spyOn(global, 'fetch')
        .mockImplementation(() => createMockFetchFailure(true));
    });

    afterAll(() => global.fetch.mockClear());

    it('fetches data from server and sets error state', done => {
      expect(global.fetch).toHaveBeenCalledTimes(1);
      expect(global.fetch).toHaveBeenCalledWith(FETCH_URL);

      process.nextTick(() => {
        expect(wrapper.find(TaskDetail).state()).toEqual(failResponse);

        done();
      });
    });

    it('does not render', () => {
      expect(wrapper.contains(<TaskDetail />)).toBe(false);
    });
  });

  describe('when the fetch succeeds', () => {
    beforeAll(() => {
      jest
        .spyOn(global, 'fetch')
        .mockImplementation(() => createMockFetchSuccess(taskDetail));
    });

    beforeEach(() => wrapper.update());

    afterAll(() => global.fetch.mockClear());

    it('fetches data from server and sets correct state', done => {
      expect(global.fetch).toHaveBeenCalledTimes(1);
      expect(global.fetch).toHaveBeenCalledWith(FETCH_URL);

      process.nextTick(() => {
        expect(wrapper.find(TaskDetail).state()).toEqual(successResponse);
        done();
      });
    });

    it('matches snapshot', () => {
      expect(toJson(wrapper.find(TaskDetail))).toMatchSnapshot();
    });
  });
});
