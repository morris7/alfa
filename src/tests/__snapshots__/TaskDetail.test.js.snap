// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`when rendering a TaskDetail component when the fetch succeeds matches snapshot 1`] = `
.c1 {
  box-shadow: 0 4px 4px 0 rgba(0,0,0,0.26),0 7px 14px 0 rgba(0,0,0,0.19);
  border-radius: 2px;
  margin-bottom: 25px;
}

.c2 {
  background: #ffffff;
  width: 100%;
  min-height: 200px;
  padding: 15px;
  margin-bottom: 25px;
}

.c2 li {
  list-style: none;
  padding: 10px 0px;
}

.c2 li img {
  padding-right: 10px;
}

.c2 li:nth-child(even) {
  background: #f2f2f2;
}

.c0 a {
  -webkit-text-decoration: none;
  text-decoration: none;
}

<TaskDetail
  match={
    Object {
      "params": Object {
        "id": 1,
      },
    }
  }
>
  <styled.article>
    <StyledComponent
      forwardedComponent={
        Object {
          "$$typeof": Symbol(react.forward_ref),
          "attrs": Array [],
          "componentStyle": ComponentStyle {
            "componentId": "sc-htpNat",
            "isStatic": true,
            "lastClassName": "c0",
            "rules": Array [
              "
  a {
    text-decoration: none;
  }
",
            ],
          },
          "displayName": "styled.article",
          "foldedComponentIds": Array [],
          "render": [Function],
          "styledComponentId": "sc-htpNat",
          "target": "article",
          "toString": [Function],
          "warnTooManyClasses": [Function],
          "withComponent": [Function],
        }
      }
      forwardedRef={null}
    >
      <article
        className="c0"
      >
        <Panel>
          <styled.div>
            <StyledComponent
              forwardedComponent={
                Object {
                  "$$typeof": Symbol(react.forward_ref),
                  "attrs": Array [],
                  "componentStyle": ComponentStyle {
                    "componentId": "sc-bdVaJa",
                    "isStatic": true,
                    "lastClassName": "c1",
                    "rules": Array [
                      "
  box-shadow: 0 4px 4px 0 rgba(0, 0, 0, 0.26), 0 7px 14px 0 rgba(0, 0, 0, 0.19);
  border-radius: 2px;
  margin-bottom: 25px;
",
                    ],
                  },
                  "displayName": "styled.div",
                  "foldedComponentIds": Array [],
                  "render": [Function],
                  "styledComponentId": "sc-bdVaJa",
                  "target": "div",
                  "toString": [Function],
                  "warnTooManyClasses": [Function],
                  "withComponent": [Function],
                }
              }
              forwardedRef={null}
            >
              <div
                className="c1"
              >
                <styled.section
                  className="taskInfo"
                >
                  <StyledComponent
                    className="taskInfo"
                    forwardedComponent={
                      Object {
                        "$$typeof": Symbol(react.forward_ref),
                        "attrs": Array [],
                        "componentStyle": ComponentStyle {
                          "componentId": "sc-bwzfXH",
                          "isStatic": true,
                          "lastClassName": "c2",
                          "rules": Array [
                            "
  background: #ffffff;
  width: 100%;
  min-height: 200px;
  padding: 15px;
  margin-bottom: 25px;

  li {
    list-style: none;
    padding: 10px 0px;
    img {
      padding-right: 10px;
    }
    &:nth-child(even) {
      background: #f2f2f2;
    }
  }
",
                          ],
                        },
                        "displayName": "styled.section",
                        "foldedComponentIds": Array [],
                        "render": [Function],
                        "styledComponentId": "sc-bwzfXH",
                        "target": "section",
                        "toString": [Function],
                        "warnTooManyClasses": [Function],
                        "withComponent": [Function],
                      }
                    }
                    forwardedRef={null}
                  >
                    <section
                      className="taskInfo c2"
                    >
                      <h2>
                        Task information
                      </h2>
                      <hr />
                      <p>
                        Task body
                      </p>
                    </section>
                  </StyledComponent>
                </styled.section>
              </div>
            </StyledComponent>
          </styled.div>
        </Panel>
        <Panel>
          <styled.div>
            <StyledComponent
              forwardedComponent={
                Object {
                  "$$typeof": Symbol(react.forward_ref),
                  "attrs": Array [],
                  "componentStyle": ComponentStyle {
                    "componentId": "sc-bdVaJa",
                    "isStatic": true,
                    "lastClassName": "c1",
                    "rules": Array [
                      "
  box-shadow: 0 4px 4px 0 rgba(0, 0, 0, 0.26), 0 7px 14px 0 rgba(0, 0, 0, 0.19);
  border-radius: 2px;
  margin-bottom: 25px;
",
                    ],
                  },
                  "displayName": "styled.div",
                  "foldedComponentIds": Array [],
                  "render": [Function],
                  "styledComponentId": "sc-bdVaJa",
                  "target": "div",
                  "toString": [Function],
                  "warnTooManyClasses": [Function],
                  "withComponent": [Function],
                }
              }
              forwardedRef={null}
            >
              <div
                className="c1"
              >
                <styled.section
                  className="logInfo"
                >
                  <StyledComponent
                    className="logInfo"
                    forwardedComponent={
                      Object {
                        "$$typeof": Symbol(react.forward_ref),
                        "attrs": Array [],
                        "componentStyle": ComponentStyle {
                          "componentId": "sc-bwzfXH",
                          "isStatic": true,
                          "lastClassName": "c2",
                          "rules": Array [
                            "
  background: #ffffff;
  width: 100%;
  min-height: 200px;
  padding: 15px;
  margin-bottom: 25px;

  li {
    list-style: none;
    padding: 10px 0px;
    img {
      padding-right: 10px;
    }
    &:nth-child(even) {
      background: #f2f2f2;
    }
  }
",
                          ],
                        },
                        "displayName": "styled.section",
                        "foldedComponentIds": Array [],
                        "render": [Function],
                        "styledComponentId": "sc-bwzfXH",
                        "target": "section",
                        "toString": [Function],
                        "warnTooManyClasses": [Function],
                        "withComponent": [Function],
                      }
                    }
                    forwardedRef={null}
                  >
                    <section
                      className="logInfo c2"
                    >
                      <h2>
                        Log 
                      </h2>
                      <hr />
                      <ul>
                        <li
                          key="0"
                        >
                          <img
                            src="phone.svg"
                          />
                          <span>
                            content
                          </span>
                        </li>
                      </ul>
                      <Button
                        text="Add Call"
                      >
                        <button>
                          Add Call
                        </button>
                      </Button>
                      <Button
                        text="Add Note"
                      >
                        <button>
                          Add Note
                        </button>
                      </Button>
                    </section>
                  </StyledComponent>
                </styled.section>
              </div>
            </StyledComponent>
          </styled.div>
        </Panel>
        <Link
          to="/"
        >
          <a
            href="/"
            onClick={[Function]}
          >
            Back
          </a>
        </Link>
      </article>
    </StyledComponent>
  </styled.article>
</TaskDetail>
`;
